### PR TITLE
fix(line): soft clipping gradient.

### PR DESCRIPTION
**Pull request by pissang**
## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others

### What does this PR do?

In the cases like #15680. The coordinates range of gradient objects is very large. Which leads to blurry transition when GPU acceleration is enabled in chrome. Only version 5.1 doesn't have this issue because we clamped the data. But it will cause infinite value wrong appearance. So in this PR https://github.com/apache/echarts/pull/15416 we removed the clamping and this blurry issue appears again.

In this PR. I clamp the color stops and recalculate the clamped color in `LineView.ts`. It should fix both issues mentioned above.

### Fixed issues

#15680 

### Before: What was the problem?

image

### After: How is it fixed in this PR?

image

## Misc

- [ ] The API has been changed (apache/echarts-doc#xxx).
- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

NA.

### Merging options

- [ ] Please squash the commits into a single one when merge.

### CODE REVIEW

**Comment by plainheart** [comment]
It can solve but doesn't seem to solve perfectly...

**Comment by pissang** [comment]
@plainheart Yes, the leaking issue is hard to be avoided because of the limitation of the gradient.

**Review by plainheart — APPROVED** [review]

File: src/chart/line/LineView.ts [hunk 1]
```diff
@@ -257,34 +300,40 @@ function getVisualGradient(
...
+        // All stops are out of range. All will be the same color.
+        return colorStops[0].coord < 0
+            ? (outerColors[1] ? outerColors[1] : colorStops[stopLen - 1].color)
+            : (outerColors[0] ? outerColors[0] : colorStops[0].color);
+    }
 
-    const tinyExtent = 10; // Arbitrary value: 10px
-    const minCoord = colorStops[0].coord - tinyExtent;
-    const maxCoord = colorStops[stopLen - 1].coord + tinyExtent;
+    const tinyExtent = 0; // Arbitrary value: 10px
```
Comment (plainheart) [inline src/chart/line/LineView.ts:314]: Is it still necessary to get involved in the calculation?

**Comment by dignfei** [comment]
大佬牛逼

File: src/chart/line/LineView.ts (hunk 1 above)
Comment (pissang) [reply src/chart/line/LineView.ts:314]: I'm not sure but I will prefer to keep it to avoid breaking in the patch version.

Comment (plainheart) [reply src/chart/line/LineView.ts:314]: OK.

**Comment by pissang** [comment]
@plainheart I did some search on how skia draws the gradient with `OpenGL`. Seems it will use a `256` width texture to cache the gradient pixels when it has more than 10 color stops. The size is much smaller than the canvas width. So it can't be pixel-accurate.

https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/src/gpu/gradients/GrGradientShader.cpp;l=29;drc=a9641800e0e8f7a2ea9a5f55f831b632e990a0e0;bpv=1;bpt=1

One way to solve this issue I can think of is to split the gradient into multiples and draw them on a full-size canvas. Then use this canvas as a CanvasPattern.

**Pull request merged**